src/Core/Shop/Url/CategoryProvider.php

```diff
@@ -0,0 +1,61 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Shop\Url;
+
+use Link;
+
+/**
+ * Class CategoryProvider provides base Front Office URL for context shop.
+ */
+final class CategoryProvider implements UrlProviderInterface
+{
+    /**
+     * @var Link
+     */
+    private $link;
+
+    /**
+     * @param Link $link
+     */
+    public function __construct(Link $link)
```

Comment:
This is a legacy class inside the core :/ If you want to do this, create an interface inside the legacy, and implement it inside the adapter namespace. It will help refactoring.

Reply:
You mentioned i shoudn't use adapter classes in Core, so I thought according to this ADR
https://github.com/PrestaShop/ADR/blob/master/0014-drop-the-no-use-of-legacy-in-new-code-restriction.md
Copy the implementation from Adapter to Core, I guess should also change the CategoryProvider in adapter class to use in core, but isn't the whole point of ADR that we can use legacy in core?